synapse/rest/client/v2_alpha/report_event.py

```diff
@@ -58,17 +94,250 @@ async def on_POST(self, request, room_id, event_id):
                 Codes.BAD_JSON,
             )
 
+        target = body.get("org.matrix.msc2938.target", "homeserver-admins")
+
+        if target == "room-moderators":
+            # Report event to room moderators.
+            # This branch has further safety checks (e.g. can the user actually see the event?)
+            return await self._report_to_room_moderators(
+                room_id=room_id,
+                event_id=event_id,
+                user_id=user_id,
+                reason=body["reason"],
+                score=body["score"],
+            )
+        elif target == "homeserver-admins":
+            return await self._report_to_homeserver_admin(
+                room_id=room_id,
+                event_id=event_id,
+                user_id=user_id,
+                reason=body["reason"],
+                content=body,
+            )
+        else:
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST,
+                "Optional param 'target' must be one of ['homeserver-admins', 'room-moderators']",
+                Codes.BAD_JSON,
+            )
+
+    async def _report_to_homeserver_admin(
+        self,
+        room_id: RoomID,
+        event_id: EventID,
+        user_id: UserID,
+        reason: str,
+        content: JsonDict,
+    ):
+        """
+        Report an event to the homeserver admin.
+        This is typically meant to be used to report an entire room, e.g. for harboring illegal activities.
+
+        - room_id The room in which the event took place.
+        - event_id The event to report. We do not check whether the event took place in that room.
+        - user_id The user who reported the event. We do not check whether the user can actually witness the event.
+        - reason The human-readable reason provided by the user.
+        - content A JSON dictionary `{reason: String?, score: Number?}`.
```

Comment:
Have a look at the rest of Synapse for conventions for documenting parameters. (As per https://github.com/matrix-org/synapse/blob/develop/docs/code_style.md#general-rules, we use the conventions at https://google.github.io/styleguide/pyguide.html#doc-function-args)